src/html/html-document.ts

```diff
@@ -102,21 +103,42 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
     };
   }
 
-  stringify() {
-    // TODO(issues/315): serialize inline documents correctly.
-    let result = parse5.serialize(this.ast);
+  stringify(options?: StringifyOptions) {
+    options = options || {};
+    // We want to make of copy of `this` and all of the inline documents such
+    // that cross-references between the asts are maintained. Fortunately,
+    // clone() does this! So we'll clone them all together.
+    const immutableDocuments = options.inlineDocuments || [];
+    immutableDocuments.unshift(this);
+    // We can modify these, as they don't escape this method.
+    const mutableDocuments = clone(immutableDocuments);
+    const self = mutableDocuments.shift();
 
-    // Strip out inferred boilerplate nodes that are injected.
-    const injectedTagNames = ['html', 'head', 'body'];
-    for (const tagName of injectedTagNames) {
-      if (!this.contents.includes(`<${tagName}`)) {
-        result =
-            result.replace(RegExp(`<${tagName}>([^]*)?</${tagName}>`), '$1');
-      }
+    for (const doc of mutableDocuments) {
+      // TODO(rictic): infer this from doc.astNode's indentation.
+      const expectedIndentation = 2;
+
+      dom5.setTextContent(
+          doc.astNode,
+          '\n' + doc.stringify({indent: expectedIndentation}) + '  '.repeat(1));
```

Comment:
I guess the thinking is that we only need newlines if the total content of this line is >80cols or whatever?

That's way more clever than I think we're likely to ever make this method, IMO.